curvefs/src/client/s3/client_s3_cache_manager.cpp

```diff
@@ -275,69 +275,97 @@ int FileCacheManager::Read(uint64_t inodeId, uint64_t offset, uint64_t length,
+                    if (s3InfoListIter == inode->s3chunkinfomap().end()) {
+                        VLOG(6)
+                            << "s3infolist is not found.index:" << iter->index;
+                        memset(dataBuf + iter->bufOffset, 0, iter->len);
+                        continue;
+                    }
+                    std::vector<S3ReadRequest> s3Requests;
+                    GenerateS3Request(*iter, s3InfoListIter->second, dataBuf,
+                                      &s3Requests, inode->fsid(),
+                                      inode->inodeid());
+                    totalS3Requests.insert(totalS3Requests.end(),
+                                           s3Requests.begin(),
+                                           s3Requests.end());
+                }
+            }
+            if (totalS3Requests.empty()) {
+                VLOG(6) << "s3 has not data to read.";
+                return readOffset;
             }
-            std::vector<S3ReadRequest> s3Requests;
-            GenerateS3Request(*iter, s3InfoListIter->second, dataBuf,
-                              &s3Requests, inode->fsid(), inode->inodeid());
-            totalS3Requests.insert(totalS3Requests.end(), s3Requests.begin(),
-                                   s3Requests.end());
-        }
-    }
-    if (totalS3Requests.empty()) {
-        VLOG(6) << "s3 has not data to read.";
-        return readOffset;
-    }
 
-    uint32_t i;
-    for (i = 0; i < totalS3Requests.size(); i++) {
-        S3ReadRequest &tmp_req = totalS3Requests[i];
-        VLOG(9) << "S3ReadRequest chunkid:" << tmp_req.chunkId
-                << ",offset:" << tmp_req.offset << ",len:" << tmp_req.len
-                << ",objectOffset:" << tmp_req.objectOffset
-                << ",readOffset:" << tmp_req.readOffset
-                << ",compaction:" << tmp_req.compaction
-                << ",fsid:" << tmp_req.fsId << ",inodeId:" << tmp_req.inodeId;
-    }
+            uint32_t i;
+            for (i = 0; i < totalS3Requests.size(); i++) {
+                S3ReadRequest& tmp_req = totalS3Requests[i];
+                VLOG(9) << "S3ReadRequest chunkid:" << tmp_req.chunkId
+                        << ",offset:" << tmp_req.offset
+                        << ",len:" << tmp_req.len
+                        << ",objectOffset:" << tmp_req.objectOffset
+                        << ",readOffset:" << tmp_req.readOffset
+                        << ",compaction:" << tmp_req.compaction
+                        << ",fsid:" << tmp_req.fsId
+                        << ",inodeId:" << tmp_req.inodeId;
+            }
 
-    std::vector<S3ReadResponse> responses;
+            std::vector<S3ReadResponse> responses;
 
-    ret = ReadFromS3(totalS3Requests, &responses, fileLen);
-    if (ret < 0) {
-        LOG(ERROR) << "read from s3 failed. ret:" << ret;
-        return ret;
-    }
+            ret = ReadFromS3(totalS3Requests, &responses, fileLen);
+            if (ret < 0) {
```

Comment:
If one request fails, retry everything?

Reply:
if obj does not exit, we assume the s3infomap cache is stale.
totalS3Requests is generated by the stale inode s3infomap cache.
we need retry the whole process, including generating totalS3Request and send request.